pkg/vsphere/adapter.go

```diff
@@ -80,41 +97,192 @@ func (a *vAdapter) Start(ctx context.Context) error {
+			if err != nil {
+				// TODO: return instead?
+				logger.Errorf("send events: success %d (total %d): %v", n, len(events), err)
+
+				// 	special case: all events failed so skipping checkpoint
+				if n == 0 {
+					continue
+				}
+			}
+
+			if n == 0 && err == nil {
+				panic("we should never get here")
 			}
+
+			// last successfully sent event from batch
+			lastEvent = events[n-1]
+			cp := checkpoint{
+				VCenter:               a.Source,
+				LastEventKey:          lastEvent.GetEvent().Key,
+				LastEventType:         getEventDetails(lastEvent).Type,
+				LastEventKeyTimestamp: lastEvent.GetEvent().CreatedTime,
+				CreatedTimestamp:      time.Now().UTC(),
+			}
+			if err = a.KVStore.Set(ctx, checkpointKey, cp); err != nil {
+				return fmt.Errorf("set checkpoint: %w", err)
+			}
+
+			bOff.Reset()
 		}
+	}
+}
+
+// sendEvents converts all events to cloud events and sends them to the
+// configured sink. It returns the number of successfully processed events,
+// which might 0, partial or all events. sendEvents returns when all events are
+// processed or on the first error.
+func (a *vAdapter) sendEvents(ctx context.Context, baseEvents []types.BaseEvent) (int, error) {
+	var success int
+
+	for _, be := range baseEvents {
+		ev := cloudevents.NewEvent(cloudevents.VersionV1)
+		ev.SetSource(a.Source)
+
+		details := getEventDetails(be)
+		ev.SetType("com.vmware.vsphere." + details.Type)
+		ev.SetExtension("EventClass", details.Class)
+
+		// TODO: ingestion time?
+		ev.SetTime(be.GetEvent().CreatedTime)
 
-		return nil
+		// TODO: UUID?
+		ev.SetID(fmt.Sprintf("%d", be.GetEvent().Key))
+
+		// TODO(mattmoor): Consider setting the subject
+
+		// TODO: make encoding configurable?
+		if err := ev.SetData(cloudevents.ApplicationXML, be); err != nil {
+			return success, fmt.Errorf("set data on event: %w", err)
+		}
+
+		// TODO: better partial batch failure handling here?
+		result := a.CEClient.Send(ctx, ev)
+		if !cloudevents.IsACK(result) {
+			logging.FromContext(ctx).Errorw("failed to send cloudevent", zap.Error(result))
+			return success, result
+		}
+		success++
+	}
+
+	return success, nil
+}
+
+// getBeginFromCheckpoint returns the valid begin time to start replaying
+// vCenter events. If the checkpoint is empty the current vCenter time (UTC) is
+// used. If the last checkpoint event timestamp is larger than maxAge, replay
+// will start at maxAge.
+func getBeginFromCheckpoint(ctx context.Context, vcTime time.Time, cp checkpoint, maxAge time.Duration) time.Time {
+	begin := vcTime
+	logger := logging.FromContext(ctx)
+
+	cpTime := cp.LastEventKeyTimestamp
+	if !cpTime.IsZero() {
+		// valid checkpoint
+		logger.Info("found existing checkpoint")
+		maxTime := begin.Add(maxAge * -1)
+		if maxTime.Unix() > cpTime.Unix() {
+			logger.Warnw("last event timestamp in checkpoint is older than configured maximum", "maxHistory", maxAge.String(), "checkpointTimestamp", cp.LastEventKeyTimestamp.String())
```

Comment:
nit, might want to include in the log message that this means some events can be lost?

Reply:
Good point, explicit now:

`logger.Warnw("potential data loss: last event timestamp in checkpoint is older than configured maximum",`